testapp/testapp/AppDelegate.swift

```diff
@@ -0,0 +1,150 @@
+            }
+        }
+        realtime = ARTRealtime(options: options)
+        realtime.connection.on { state in
+            if let state = state {
+                switch state.current {
+                case .connected:
+                    print("connected")
+                case .failed:
+                    print("failed")
+                default:
+                    break
+                }
+            }
+        }
+        return realtime
+    }
+    func getTokenRequest(completion: @escaping (NSDictionary?, Error?) -> ())  {
+        let requestURL = URL(string: authURL)!
+        let urlRequest = URLRequest(url: requestURL as URL)
+        let session = URLSession.shared
+        let task = session.dataTask(with: urlRequest) {
+            (data, response, error) -> Void in
+            let httpResponse = response as! HTTPURLResponse
+            let statusCode = httpResponse.statusCode
+            if (statusCode == 200) {
+                do{
+                    let json = try JSONSerialization
+                        .jsonObject(with: data!, options:.allowFragments) as! NSDictionary
+                    completion(json, nil)
+                } catch {
+                    print("There was an error while obtaining JSON")
+                }
+            }
+        }
+        task.resume()
+    }
+
+    func didDeactivateAblyPush(_ error: ARTErrorInfo?) {
+        if let error = error {
+            // Handle error
+            print("** push de-activation failed", error)
+            return
+        }
+        print("** push de-activated, re-activating")
+        self.realtime.push.activate()
+    }
+
+    func didActivateAblyPush(_ error: ARTErrorInfo?) {
+        if let error = error {
+            // Handle error
+            print("** push activation failed, err=\(String(describing: error))")
+            return
+        }
+        print("** push activated")
+    }
+
+    func ablyPushCustomRegister(_ error: ARTErrorInfo?, deviceDetails: ARTDeviceDetails?, callback: @escaping (ARTDeviceIdentityTokenDetails?, ARTErrorInfo?) -> Void) {
+        if error != nil {
+            // Handle error
+            print ("** failed to custom register: \(String(describing: error))")
+        }
+        let request = NSMutableURLRequest(url: NSURL(string: "\(subscribeURL)?deviceId=\(deviceDetails?.id ?? "")&deviceToken=\(deviceTokenString ?? "")")! as URL,
+                                          cachePolicy: .useProtocolCachePolicy,
+                                          timeoutInterval: 10.0)
+        request.httpMethod = "GET"
+        let session = URLSession.shared
+        let dataTask = session.dataTask(with: request as URLRequest, completionHandler: { (data, response, error) -> Void in
+            if (error != nil) {
+                print(error)
+            } else {
+                let httpResponse = response as? HTTPURLResponse
+                print(httpResponse)
+            }
+        })
+        dataTask.resume()
```

Comment:
@cjbatin Is this working?! You're not calling the `callback`.

Reply:
Yeah it seems to work fine. I was unsure as well, I have no idea what the callback is for.

Reply:
The `callback` should be called when your custom device registration has finished processing. You must execute this closure at some point after processing the server response to let the library know that you're done. Maybe I should add this explanation to the method description and change the `callback` attribute name to `completion`.

Reply:
@Srushtika This would mean changing the server callback I believe. However it all works fine without this so I'm still not 100% sure why it's required?

Reply:
I guess I could edit it to just return nil for both cases?

Reply:
@cjbatin Let me first check why it's working as usual. Something should be misplaced. Thanks 👍

Reply:
@cjbatin I tested this and the `callback` needs to be called. From the log, you can see that the Activation State Machine is blocked, waiting for the device registration:

```
INFO : KSCrash.m (346): -[KSCrash sendAllReportsWithCompletion:]: Sending 0 crash reports
ClientId C03BC116-8004-4D78-A71F-8CA3122734DB
2019-02-21 14:18:11.069512+0000 AblyPushSwift[11129:3077328] handling event ARTPushActivationEventCalledActivate from ARTPushActivationStateNotActivated
2019-02-21 14:18:11.069614+0000 AblyPushSwift[11129:3077328] ARTPushActivationStateNotActivated state: transitioning to ARTPushActivationEventCalledActivate event
2019-02-21 14:18:11.172111+0000 AblyPushSwift[11129:3077328] transition: ARTPushActivationStateNotActivated -> ARTPushActivationStateWaitingForPushDeviceDetails
2019-02-21 14:18:11.224109+0000 AblyPushSwift[11129:3077221] ARTPush: device token received 3fb96bce14d2fa8e581d66be0e14b7ca746b930e434dc2024635501c5bf07c71
2019-02-21 14:18:11.224781+0000 AblyPushSwift[11129:3077221] ARTPush: device token stored
2019-02-21 14:18:11.224858+0000 AblyPushSwift[11129:3077327] handling event ARTPushActivationEventGotPushDeviceDetails from ARTPushActivationStateWaitingForPushDeviceDetails
2019-02-21 14:18:11.224892+0000 AblyPushSwift[11129:3077327] ARTPushActivationStateWaitingForPushDeviceDetails state: transitioning to ARTPushActivationEventGotPushDeviceDetails event
2019-02-21 14:18:11.224959+0000 AblyPushSwift[11129:3077327] transition: ARTPushActivationStateWaitingForPushDeviceDetails -> ARTPushActivationStateWaitingForDeviceRegistration
```

So the `func didActivateAblyPush(_ error: ARTErrorInfo?)` will never be called. The `callback` will continue with the activation flow. You must return a valid value in the `callback`. The `/push/deviceRegistrations` called in you server returns the valid device identity token.

/cc @Srushtika